### PR TITLE
Feature/version2.0

### PR DESCRIPTION
**Pull request by Juhani-moilanen**
Changes:
-  "Statistics" -- run once per day
- Run updates_log once an hour
-  Remove Settings ( diff output is the default, ultimate control removed)
- Add module version to diff mode
- "Test mode" ( Able to run regardless of "last run" time)
- Refactoring

Tests have been Refactored to be compatible with Drupal's testing system

ToDo => Have not Updated documentation 
WE Should consider removing or fixing lando to be able to run/test this module ( How is this done in other Drupal modules ?)
currently, I Am using [drupal-contributions](https://github.com/lando/drupal-contributions) for this purpose

### CODE REVIEW

File: src/UpdatesLog.php [hunk 1]
```diff
@@ -288,34 +236,83 @@ public function statusesGet(): array {
...
+  public function generateStatistics(array $statuses): array {
+    $statistics = [
+      "updates_log" => "2.0",
+      "last_check_epoch" => $this->lastUpdated,
+      "last_check_human" => gmdate('Y-m-d\Th:i:sZT', $this->lastUpdated),
+      "last_check_ago" => time() - $this->lastUpdated,
+      "summary" => [
+        "CURRENT" => 0,
+        "NOT_CURRENT" => 0,
+        "OUTDATED" => 0,
```
Comment (ragnarkurmwunder) [inline src/UpdatesLog.php:270]: "OUTDATED" and "NOT_CURRENT" the same thing?
or accidentally forgot "NOT_CURRENT"?

File: src/UpdatesLog.php [hunk 2]
```diff
@@ -275,10 +224,9 @@ public function statusesGet(): array {
       -4 => 'FETCH_PENDING',
```
Comment (ragnarkurmwunder) [inline src/UpdatesLog.php:224]: shouldn't we somehow handle OUTDATED?

File: tests/src/Kernel/LogTest.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+
+    $this->installConfig(['updates_log']);
+    $this->installSchema('dblog', ['watchdog']);
+
+
```
Comment (ragnarkurmwunder) [inline tests/src/Kernel/LogTest.php:39]: extra newlines

File: tests/src/Kernel/LogTest.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    /** @var \Drupal\updates_log\UpdatesLog $service */
+    $this->service = \Drupal::service('updates_log.updates_logger');
+    $this->db = \Drupal::database();
+
```
Comment (ragnarkurmwunder) [inline tests/src/Kernel/LogTest.php:43]: extra newline

File: tests/src/Kernel/LogTest.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\updates_log\UpdatesLog;
+
+/**
+ * @coversDefaultClass \Drupal\updates_log\UpdatesLog
+ * @group updates_log
+ */
+class LogTest extends KernelTestBase {
+
+  private UpdatesLog $service;
```
Comment (ragnarkurmwunder) [inline tests/src/Kernel/LogTest.php:15]: could it be `$updatesLogService`?
easier to read.

File: tests/src/Kernel/LogTest.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    $time = time();
+    $statistics = [
+      "updates_log" => "2.0",
+      "last_check_epoch" => $time,
+      "last_check_human" => gmdate('Y-m-d\Th:i:sZT', $time),
+      "last_check_ago" => 1,
+      "summary" => [
+        "CURRENT" => 2,
+        "NOT_CURRENT" => 1,
+        "OUTDATED" => 1,
```
Comment (ragnarkurmwunder) [inline tests/src/Kernel/LogTest.php:79]: outdated vs not_current

File: tests/src/Kernel/RunTest.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\updates_log\UpdatesLog;
+
+/**
+ *
+ * @group updates_log
+ */
+class RunTest extends KernelTestBase {
+
+  private UpdatesLog $service;
```
Comment (ragnarkurmwunder) [inline tests/src/Kernel/RunTest.php:14]: var name, see the other comment

File: tests/src/Kernel/RunTest.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+
+    $this->installConfig(['updates_log']);
+    /** @var \Drupal\updates_log\UpdatesLog $service */
+    $this->service = \Drupal::service('updates_log.updates_logger');
+
```
Comment (ragnarkurmwunder) [inline tests/src/Kernel/RunTest.php:35]: extra newline

File: tests/src/Kernel/StatusesGetTest.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+
+    $this->installConfig(['updates_log']);
+    /** @var \Drupal\updates_log\UpdatesLog $service */
+    $this->service = \Drupal::service('updates_log.updates_logger');
+
```
Comment (ragnarkurmwunder) [inline tests/src/Kernel/StatusesGetTest.php:36]: extra nl

File: tests/src/Unit/ComputeDiffTest.php [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+   * @covers ::computeDiff
+   */
+  public function testComputeDiffSame(): void {
+    $int = $this->updatesLog->computeDiff(
+      ['x' => ['status' => 'x', 'version' => 'x']],
+      ['x' => 'x']
+    );
+    $this->assertSame([], $int);
+
+
```
Comment (ragnarkurmwunder) [inline tests/src/Unit/ComputeDiffTest.php:23]: nl

File: tests/src/Unit/GenerateStatisticsTest.php [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        'b' => ['status' => 'CURRENT', 'version' => 'x'],
+        'c' => ['status' => '???', 'version' => 'x'],
+
+      ]
+    );
+    $this->assertEquals(2, $statistics['summary']['CURRENT']);
+    $this->assertEquals(1, $statistics['summary']['NOT_CURRENT']);
+    $this->assertEquals(1, $statistics['summary']['UNKNOWN']);
+    $this->assertCount(4, $statistics['details']);
+
```
Comment (ragnarkurmwunder) [inline tests/src/Unit/GenerateStatisticsTest.php:32]: nl

File: tests/src/Unit/ShouldUpdateTest.php [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  public function testLater(): void {
+    $status = $this->updatesLog->shouldUpdate(time(), time() - 24 * 60 * 60);
+    $this->assertTrue($status);
+  }
+
+  /**
+   * @covers ::ShouldUpdate
+   */
+  public function testTestEnv(): void {
+    putenv('UPDATES_LOG_TEST=1');
```
Comment (ragnarkurmwunder) [inline tests/src/Unit/ShouldUpdateTest.php:41]: unrelated.
we could consider setting up the project for Silta too:
* to run code scanners
* to run tests

File: tests/src/Unit/UpdatesLogTestBase.php [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+ * @group updates_log
+ */
+class UpdatesLogTestBase extends UnitTestCase {
+
+  protected UpdatesLog $updatesLog;
+
+  protected State $state;
+
+
+  protected function setUp(): void {
```
Comment (ragnarkurmwunder) [inline tests/src/Unit/UpdatesLogTestBase.php:27]: can you add a bit more comments here?
it's half Chinese :)

File: tests/src/Unit/UpdatesLogTestBase.php [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+use Drupal\updates_log\UpdatesLog;
+use Prophecy\Argument;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * @coversDefaultClass \Drupal\updates_log\UpdatesLog
+ * @group updates_log
+ */
+class UpdatesLogTestBase extends UnitTestCase {
```
Comment (ragnarkurmwunder) [inline tests/src/Unit/UpdatesLogTestBase.php:20]: by the way, we can test private functions too.
see in the guild channel.

**Review by ragnarkurmwunder — CHANGES_REQUESTED** [review]

File: README.md [hunk 15]
```diff
@@ -62,14 +62,6 @@ In `settings.php` add the following:
 $config['updates_log']['diff'] = TRUE;
 ```
 
-### Control the scheduling
```
Comment (ragnarkurmwunder) [inline README.md:65]: We should add info how to test 2.0.
How to force immediate output?
How to get statistical output?

**Review by ragnarkurmwunder — CHANGES_REQUESTED** [review]

File: src/UpdatesLog.php [hunk 16]
```diff
@@ -275,10 +224,9 @@ public function statusesGet(): array {
...
     ];
 
-    /** @var array<mixed> */
     $available = update_get_available(TRUE);
 
-    /** @var array<string, array{status: int}> */
+    /** @var array<string, array{status: int}> $available */
```
[the remark this answers is not in the document]
Comment (Juhani-moilanen) [reply src/UpdatesLog.php:263]: > When documenting variables with the @var tag note that you should use the special /** */ delimiters with a double opening asterisk. Always use the fully qualified namespace of the class or interface, and include the name of the variable at the end.


https://www.drupal.org/docs/develop/standards/php/api-documentation-and-comment-standards

File: src/UpdatesLog.php (hunk 1 above)
Comment (Juhani-moilanen) [reply src/UpdatesLog.php]: Seems That OUTDATED somehow went in there instead of NOT_CURRENT
is fixed now

**Comment by Juhani-moilanen** [comment]
Fixed all code issues found by `lando grumphp run`

File: src/UpdatesLog.php [hunk 17]
```diff
@@ -82,29 +115,23 @@ public function computeDiff(array $new, array $old): array {
 
     $diff = [];
 
-    foreach ($new as $project => $status) {
+    foreach ($new as $project => $data) {
```
[the remark this answers is not in the document]
Comment (Juhani-moilanen) [reply src/UpdatesLog.php:149]: ATM we do not save the "version" in State so the "old_statuses" structure is the same 

Should we save the version?

**Pull request merged**